MREUnityRuntime/MREUnityRuntimeLib/Core/Physics/JitterBuffer.cs

```diff
@@ -232,235 +320,577 @@ public void add(float bufferTime, float meanBufferTime, float targetBufferTime,
+#endregion
+
+#region Time Management
+
+		/// <summary>
+		/// Time precision is 1 millisecond.
+		/// </summary>
+		private const float _timePrecision = 0.001f;
+
+		/// <summary>
+		/// todo: add comment
+		/// </summary>
+		private float _speedUpCoef = 0.2f;
+
+		/// <summary>
+		/// todo: add comment
+		/// </summary>
+		private float _speedDownCoef = 0.5f;
+
+		private float caclNextTimeStamp(float timestep)
+		{
+			if (CurrentLocalTime < 0.0f)
 			{
-				Id = id;
-				CurrentSnapshot = null;
-				CurrentLocalTime = float.MinValue;
+				return LastSnapshotLocalTime;
 			}
 
-			public void addSnapshot(Snapshot snapshot)
+			const float bufferTimeBiasTimeUnit = 1.0f / 60;
+
+			float targetBufferTime = _runningStats.Mean - _runningStats.Deviation;
+
+			float biasedTargetBufferTime = targetBufferTime / bufferTimeBiasTimeUnit;
+			biasedTargetBufferTime = targetBufferTime > 0 ?
+				((int)biasedTargetBufferTime) * bufferTimeBiasTimeUnit : ((int)biasedTargetBufferTime - 1) * bufferTimeBiasTimeUnit;
+
+			float nextTimestamp = CurrentLocalTime + timestep;
+			float bufferedTime = LastSnapshotLocalTime - nextTimestamp;
+
+			if (bufferedTime > 1.0f)
 			{
-				SnapshotBuffer.addSnapshot(snapshot);
+				int x = 0;
+				x++;
 			}
 
-			public void step(float timestep)
+			_runningStats.AddSample(bufferedTime);
+
+			// check if time shift is required
+			float timeShift;
+			if (Math.Abs(biasedTargetBufferTime) > 0.001)
+			{
+				// todo: limit slow-down, don't allow time to move to the past
+				timeShift = biasedTargetBufferTime > 0 ?
+					_speedUpCoef * biasedTargetBufferTime : _speedDownCoef * biasedTargetBufferTime;
+
+				nextTimestamp += timeShift;
+				_runningStats.ShiftTime(-timeShift);
+			}
+			else
 			{
-				if (_mode == Mode.Init)
+				timeShift = 0;
+			}
+
+#if DEBUG_JITTER_BUFFER
+					_debugStats.add(bufferedTime, _runningStats.Mean, targetBufferTime, biasedTargetBufferTime, timeShift, nextTimestamp);
+#endif
+
+			return nextTimestamp;
+		}
+
+#endregion
+
+#region Buffer Management
+
+		/// <summary>
+		/// Get previous and next snapshot for specified timestamp.
+		/// Snapshots older than timestamp will be deleted.
+		/// </summary>
+		//private void findSnapshots(float time, out Snapshot previous, out Snapshot nextOrCurrent)
+		//{
+		//	UnityEngine.Debug.Assert(_snapshots.Count == 0 || time <= _snapshots.Count);
+
+		//	// Find appropriate snapshots
+		//	int index = 0;
+		//	while (index < _snapshots.Count && time - _snapshots.Keys[index] > _timePrecision) index++;
+
+		//	previous = index == 0 ? null : _snapshots.Values[index - 1];
+		//	nextOrCurrent = index < _snapshots.Count ? _snapshots.Values[index] : null;
+
+		//	// Remove old snapshots
+		//	if (nextOrCurrent != null)
+		//	{
+		//		int removeThreshold = time - nextOrCurrent.Time >= -_timePrecision ? index : index - 1;
+		//		for (int r = 0; r < removeThreshold; r++) _snapshots.RemoveAt(0);
+		//	}
+		//	else
+		//	{
+		//		_snapshots.Clear();
+		//	}
+		//}
```

Comment:
Is this still needed for something coming?  Remove if stale?

Reply:
I'll delete it

---
In reply to: [450355608](https://github.com/microsoft/mixed-reality-extension-unity/pull/210#discussion_r450355608) [](ancestors = 450355608)